pyogrio/raw.py

```diff
@@ -84,19 +85,36 @@ def read(
             "geometry": "<geometry type>"
         }
     """
+    from_buffer = False
+    if isinstance(path_or_buffer, bytes):
+        from_buffer = True
+        ext = ""
+        if path_or_buffer[:4].startswith(b'PK\x03\x04'):
+            ext = ".zip"
+        path = buffer_to_virtual_file(path_or_buffer, ext=ext)
+        if path_or_buffer[:4].startswith(b'PK\x03\x04'):
+            path = "/vsizip/" + path
```

Comment:
This is now also a bit custom, but could fall under https://github.com/geopandas/pyogrio/issues/21 for more general support of paths that are converted to vsi paths under the hood (i.e. fiona's [`parse_path`/`vsi_path`](https://github.com/Toblerity/Fiona/blob/885dbd11c5dcb2c61862faefc28bfbbff99954de/fiona/path.py))

Reply:
Yeah, this seems like doing it under #21 would be better; keeps this PR more focused on in-memory inputs.

Reply:
> this seems like doing it under https://github.com/geopandas/pyogrio/issues/21 would be better; keeps this PR more focused on in-memory inputs.

The thing is that this special case added here _is_ actually for in-memory inputs. #21 is about handling string file paths / URIs (like "zip://...") and converting them into vsi paths (like "/vsizip/..."). But here we don't start with a string path as input. So addressing #21 would not cover reading an in-memory zipped file.